vervet-underground/metrics.go

```diff
@@ -0,0 +1,47 @@
+package main
+
+import (
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// VU server metrics.
+var httpDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+	Name: "vu_http_response_time_seconds",
+	Help: "Duration of HTTP requests.",
+}, []string{"path", "method", "status"})
+
+// responseWriterInterceptor is a wrapper to intercept status code from ResponseWriter.
+type responseWriter struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+// WriteHeader writes status code to the http.ResponseWriter, as well as intercepting the status code for use within
+// the middleware.
+func (w *responseWriter) WriteHeader(statusCode int) {
+	w.statusCode = statusCode
+	w.ResponseWriter.WriteHeader(statusCode)
```

Comment:
```suggestion
	w.ResponseWriter.WriteHeader(statusCode)
	if w.statusCode == 0 {
		w.StatusCode = statusCode
	}
```
The reason for inverting the calls is that `w.ResponseWriter.WriteHeader` panics is the. status code is invalid which can be recovered and handled again by the business logic.

The reason for having a conditional is that `w.ResponseWriter.WriterHeader` only actions the first valid call to it so a status code can only be set once.

Reply:
this is now removed